### PR TITLE
feat(core): custom tools, toolkits, and extension tools for tool router sessions

### PR DESCRIPTION
**Pull request by dhawal1**
## Summary

Add user-defined tools that execute locally (in-process) alongside Composio's remote tools in tool router sessions. Aligned with backend PR [hermes#8614](https://github.com/ComposioHQ/hermes/pull/8614).

**Three ways to define custom tools:**

1. **Standalone** — no auth, no toolkit association → `LOCAL_<SLUG>`
2. **Extension** — extends a Composio toolkit, inherits its auth → `LOCAL_<TOOLKIT>_<SLUG>`
3. **Custom toolkit** — groups related no-auth tools → `LOCAL_<TOOLKIT>_<SLUG>`

**Key capabilities:**
- `experimental_createTool(slug, opts)` and `experimental_createToolkit(slug, opts)` factory functions
- Tools indexed by backend for semantic search (BM25), returned to LLM via `COMPOSIO_SEARCH_TOOLS`
- `session.execute()` routes custom tools in-process, remote tools to backend — accepts original slug, `LOCAL_`-prefixed slug, or lowercase
- `COMPOSIO_MULTI_EXECUTE_TOOL` interception: splits custom/remote tools, executes in parallel, merges results
- `SessionContext` injected at runtime: `userId`, `execute()` (with sibling local routing), `proxyExecute()`
- Slug validation: alphanumeric/underscores/hyphens, no `LOCAL_` prefix, max 60 chars prefixed

### Standalone tool (no auth)

```typescript
import { experimental_createTool } from '@composio/core';
import { z } from 'zod/v3';

const getUserDetails = experimental_createTool('GET_USER_DETAILS', {
  name: 'Get user details',
  description: 'Retrieve user profile by ID',
  inputParams: z.object({ user_id: z.string() }),
  execute: async (input) => {
    return { id: input.user_id, name: 'Alice', role: 'admin' };
  },
});
```

### Extension tool (inherits Composio toolkit auth)

```typescript
const formatEmail = experimental_createTool('FORMAT_GMAIL_EMAIL', {
  name: 'Format Gmail email',
  description: 'Format an email draft with HTML formatting',
  extendsToolkit: 'gmail',
  inputParams: z.object({ to: z.string(), subject: z.string(), body: z.string() }),
  execute: async (input, session) => {
    return { formatted: true, to: input.to, html_body: input.body.replace(/\n/g, '<br>'), user_id: session.userId };
  },
});
```

### Custom toolkit (groups no-auth tools)

```typescript
import { experimental_createToolkit } from '@composio/core';

const setUserRole = experimental_createTool('SET_USER_ROLE', {
  name: 'Set user role',
  description: 'Change user role',
  inputParams: z.object({ user_id: z.string(), role: z.enum(['admin', 'developer', 'viewer']) }),
  execute: async (input) => ({ user_id: input.user_id, new_role: input.role }),
});

const userManagement = experimental_createToolkit('USER_MANAGEMENT', {
  name: 'User Management',
  description: 'Tools for managing user profiles and roles',
  tools: [setUserRole],
});
```

### Creating a session

```typescript
const composio = new Composio({ apiKey, provider: new OpenAIAgentsProvider() });

const session = await composio.create('user_123', {
  toolkits: ['weathermap', 'gmail'],
  experimental: {
    customTools: [getUserDetails, formatEmail],
    customToolkits: [userManagement],
  },
});
```

### Executing tools

```typescript
// Custom tool — runs in-process
const r1 = await session.execute('GET_USER_DETAILS', { user_id: 'user-1' });
const r2 = await session.execute('LOCAL_GET_USER_DETAILS', { user_id: 'user-1' }); // prefixed
const r3 = await session.execute('get_user_details', { user_id: 'user-1' }); // case-insensitive

// Remote tool — identical API
const weather = await session.execute('WEATHERMAP_WEATHER', { location: 'Tokyo' });

// Agent flow — SDK intercepts MULTI_EXECUTE, routes custom/remote in parallel
const tools = await session.tools();
const agent = new Agent({ tools, model: 'gpt-5.2' });
```

## Backend API alignment

Payload sent under `experimental`:
- `custom_tools`: array of `{ slug, name, description, input_schema, output_schema?, extends_toolkit? }`
- `custom_toolkits`: array of `{ slug, name, description, tools: [{ slug, name, description, input_schema }] }`

Standalone tools omit `extends_toolkit`. Extension tools set `extends_toolkit` to a valid Composio toolkit slug.

## Files changed

| File | Change |
|---|---|
| `types/customTool.types.ts` | `CustomToolSlugSchema`, `CreateCustomToolBaseSchema` (slug as 1st arg), `extendsToolkit`, `outputParams`, `CustomToolkit`, `CustomToolkitDefinition` |
| `models/CustomTool.ts` | `createCustomTool(slug, opts)`, `createCustomToolkit(slug, opts)`, `buildCustomToolsMap(tools, toolkits?)`, `serializeCustomTools`, `serializeCustomToolkits` |
| `types/toolRouter.types.ts` | `customTools`/`customToolkits` under `experimental`, `SessionProxyExecuteParams`, `proxyExecute` on Session |
| `models/ToolRouter.ts` | Read from `experimental.customTools/customToolkits`, unified `experimental` payload |
| `models/ToolRouterSession.ts` | Multi-execute interception, `session.execute()` routing, `proxyExecute()` |
| `models/SessionContext.ts` | `proxyExecute()` implementation, sibling local tool routing |
| `models/customToolExecution.ts` | Extracted `findCustomTool()`, `executeCustomTool()` |
| `models/proxyParamsTransform.ts` | Transform SDK proxy params to backend format |
| `index.ts` | `experimental_createTool`, `experimental_createToolkit` exports |
| `experimental/index.ts` | `createCustomToolkit`, `CustomToolkit`, `CreateCustomToolkitParams` |
| `docs/api/tool-router.md` | Custom tools section, `execute`/`search`/`proxyExecute` docs |
| `examples/tool-router/src/custom-tools-agent.ts` | Full agent example with all 3 tool types |

### Unit tests (79 across 2 files)

- `createCustomTool()`: slug validation (regex, no LOCAL_, invalid chars), `extendsToolkit`, `outputParams`, Zod to JSON Schema, execute signatures
- `createCustomToolkit()`: validation, rejects tools with `extendsToolkit`
- `buildCustomToolsMap()`: 3 prefix patterns, length validation, cross-group collision detection
- `serializeCustomTools()`: `extends_toolkit` mapping, omit for standalone, `output_schema`
- `serializeCustomToolkits()`: nested tool serialization
- `SessionContextImpl`: userId, remote delegation, proxyExecute, sibling routing
- `ToolRouter.create()`: payload under `experimental`, `extends_toolkit`, omit when empty
- `session.execute()`: original/prefixed slug, case-insensitive, remote fallback, error handling, Zod defaults
- `MULTI_EXECUTE split/merge`: all-local/all-remote/mixed batches, parallel execution, error merging, edge cases

### E2E tests

- Node.js fixture (`ts/e2e-tests/runtimes/node/custom-tools/`) against live API
- Focused E2E script: 18 tests covering all 3 tool types, search indexing, state mutations, error handling
- Agent example (`ts/examples/tool-router/src/custom-tools-agent.ts`) verified against staging backend — standalone + extension + custom toolkit + remote tools in a single agent run

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: ts/packages/core/src/models/CustomTool.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+ *   execute: async (input, session) => {
+ *     const result = await session.execute('META_ADS_GET_AD_ACCOUNTS', {
+ *       fields: input.fields,
+ *     });
+ *     return result.data;
+ *   },
+ * });
+ * ```
+ */
+export function CustomTool<T extends z.ZodType>(
```
Comment (haxzie) [inline ts/packages/core/src/models/CustomTool.ts]: Ideally use camel case here instead of title case. `customTool` or `createCustomTool`.

File: ts/packages/core/src/models/CustomTool.ts [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  options: NewCustomToolOptions<T>
+): CustomToolHandle {
+  const { slug, name, description, inputParams, execute, connectedToolkit } = options;
+
+  if (!slug) {
```
Comment (haxzie) [inline ts/packages/core/src/models/CustomTool.ts]: Define the tool params using zod schema, include validation and parse it here instead of individual checks. Also export the type as `CreateCustomToolParams`.

File: ts/packages/core/src/models/CustomTool.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+ *       fields: input.fields,
+ *     });
+ *     return result.data;
+ *   },
+ * });
+ * ```
+ */
+export function CustomTool<T extends z.ZodType>(
+  options: NewCustomToolOptions<T>
+): CustomToolHandle {
```
Comment (haxzie) [inline ts/packages/core/src/models/CustomTool.ts]: What does handle mean here? Can we simply call it `CustomTool`?

File: ts/packages/core/src/models/CustomTool.ts [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+ *     const result = await session.execute('META_ADS_GET_AD_ACCOUNTS', {
+ *       fields: input.fields,
+ *     });
+ *     return result.data;
+ *   },
+ * });
+ * ```
+ */
+export function CustomTool<T extends z.ZodType>(
+  options: NewCustomToolOptions<T>
```
Comment (haxzie) [inline ts/packages/core/src/models/CustomTool.ts]: Define the options as Schema and export the infered type along with the zod schema, CustomToolOptions or CustomToolParams.

File: ts/packages/core/src/models/CustomTool.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+  }
+  if (!inputParams) {
+    throw new Error('CustomTool: inputParams is required');
+  }
+  if (typeof execute !== 'function') {
+    throw new Error('CustomTool: execute must be a function');
+  }
+
+  // Convert Zod schema → JSON Schema (reuses same pattern as legacy CustomTools.createTool)
+  const paramsSchema = zodToJsonSchema.default(inputParams, {
```
Comment (haxzie) [inline ts/packages/core/src/models/CustomTool.ts:148]: Can we not discard the zod schema here? Ideally return the input params as zod schema itself so that it can be used for input validation.

File: ts/packages/core/src/models/CustomTool.ts [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+ */
+export function serializeLocalTools(
+  handles: CustomToolHandle[]
+): LocalToolDefinition[] {
+  return handles.map(handle => ({
+    slug: handle.slug,
+    name: handle.name,
+    description: handle.description,
+    input_schema: handle.inputSchema,
+    ...(handle.connectedToolkit ? { toolkit: handle.connectedToolkit } : {}),
```
Comment (haxzie) [inline ts/packages/core/src/models/CustomTool.ts]: Can we just call this variable toolkit instead of connectedToolkit?

File: ts/packages/core/src/models/ToolRouterSession.ts [hunk 7]
```diff
@@ -51,17 +62,141 @@ export class ToolRouterSession<
...
+  /**
+   * Get a tool for executing local custom tools individually.
+   * Use this for MCP flows or when you need explicit control over local tool execution.
+   *
+   * Not included in `session.tools()` — must be explicitly added to the agent's tool set.
+   *
+   * @example
+   * ```typescript
+   * const tools = await session.tools();
+   * const localTool = await session.localTools();
```
Comment (haxzie) [inline ts/packages/core/src/models/ToolRouterSession.ts]: Can we simply inject this tool automatically if customTools are registered?

File: ts/packages/core/src/models/ToolRouterSession.ts [hunk 8]
```diff
@@ -168,35 +289,206 @@ export class ToolRouterSession<
...
+    // Remote execution
     const response = await this.client.toolRouter.session.execute(this.sessionId, {
       tool_slug: toolSlug,
       arguments: arguments_ ?? {},
     });
     const transformed = transformExecuteResponse(response);
     return ToolRouterSessionExecuteResponseSchema.parse(transformed);
   }
+
+  // ── Private helpers ──────────────────────────────────────────
```
Comment (haxzie) [inline ts/packages/core/src/models/ToolRouterSession.ts:343]: Move most of these custom tool specific items to a separate file.

File: ts/packages/core/src/types/customTool.types.ts [hunk 9]
```diff
@@ -53,3 +57,99 @@ export interface ExecuteMetadata {
...
+  name: string;
+  /** Description — required for search indexing */
+  description: string;
+  /** Zod schema for input parameters */
+  inputParams: T;
+  /**
+   * Composio toolkit slug requiring an active connection (e.g. 'meta_ads').
+   * If not provided, the tool does not need Composio auth.
+   */
+  connectedToolkit?: string;
```
Comment (haxzie) [inline ts/packages/core/src/types/customTool.types.ts]: Can we call it toolkit instead of connectedToolkit?

**Review by haxzie — CHANGES_REQUESTED** [review]
What's the difference between customTool and localTool? If they're the same can we unify the names across?

File: ts/packages/core/src/models/CustomTool.ts (hunk 4 above)
Comment (dhawal1) [reply ts/packages/core/src/models/CustomTool.ts]: Done — options validated via `CreateCustomToolBaseSchema` (exported Zod schema). The full params type is `CreateCustomToolParams<T>` which extends the Zod-inferred type with `inputParams` and `execute`.

File: ts/packages/core/src/models/ToolRouterSession.ts (hunk 8 above)
Comment (dhawal1) [reply ts/packages/core/src/models/ToolRouterSession.ts:343]: Will address in a follow-up — extracting `executeLocalTool` and `findLocalTool` into standalone functions. The `routeMultiExecute` orchestration stays in ToolRouterSession since it coordinates both local and remote execution.

File: ts/packages/core/src/models/CustomTool.ts (hunk 1 above)
Comment (dhawal1) [reply ts/packages/core/src/models/CustomTool.ts]: Done — renamed to `createCustomTool()`.

File: ts/packages/core/src/types/customTool.types.ts (hunk 9 above)
Comment (dhawal1) [reply ts/packages/core/src/types/customTool.types.ts]: Done — renamed to `toolkit`.

File: ts/packages/core/src/models/CustomTool.ts (hunk 5 above)
Comment (dhawal1) [reply ts/packages/core/src/models/CustomTool.ts:148]: The Zod schema is not discarded — it's stored on the returned `CustomTool` as `inputParams` and used for runtime validation via `safeParse()` in `executeLocalTool()` (ToolRouterSession.ts:461). The `inputSchema` (JSON Schema) is only used for backend BM25 indexing.

File: ts/packages/core/src/models/ToolRouterSession.ts (hunk 7 above)
Comment (dhawal1) [reply ts/packages/core/src/models/ToolRouterSession.ts]: This is intentionally a separate `localTools()` call — it's only needed for MCP flows. `session.tools()` already handles local tools via MULTI_EXECUTE interception for SDK/agent usage. Auto-injecting the dispatcher into `tools()` would add a tool the agent doesn't need in the standard flow.

File: ts/packages/core/src/models/CustomTool.ts (hunk 3 above)
Comment (dhawal1) [reply ts/packages/core/src/models/CustomTool.ts]: Done — return type is now `CustomTool` (interface). Factory function is `createCustomTool()`.

File: ts/packages/core/src/models/CustomTool.ts (hunk 6 above)
Comment (dhawal1) [reply ts/packages/core/src/models/CustomTool.ts]: Done — renamed to `toolkit` everywhere (options, type, serialization).

File: ts/packages/core/src/models/CustomTool.ts (hunk 2 above)
Comment (dhawal1) [reply ts/packages/core/src/models/CustomTool.ts]: Done — string fields (slug, name, description, toolkit) are now validated via `CreateCustomToolBaseSchema` (Zod). `inputParams` and `execute` are checked manually since Zod can't meaningfully validate a ZodType instance or a function.

File: ts/packages/core/src/models/SessionContext.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        tool_slug: toolSlug,
+        arguments: arguments_,
+      }
+    );
+    return {
+      data: response.data,
+      error: response.error,
+      successful: !response.error,
+    };
+  }
```
[the remark this answers is not in the document]
Comment (dhawal1) [reply ts/packages/core/src/models/SessionContext.ts:62]: Fixed. `SessionContextImpl` now accepts an optional `tryLocalExecute` callback. When provided, `execute()` checks sibling local tools first (in-process) before falling back to the backend API.

The wiring happens in `ToolRouterSession.buildSessionContext()` — same `findLocalTool` + `executeLocalTool` functions used by the multi-execute router.

Also extracted `findLocalTool()` and `executeLocalTool()` into standalone functions in `localToolExecution.ts` as part of this, since they're now shared between `ToolRouterSession` and `SessionContextImpl`.

File: ts/packages/core/src/models/ToolRouterSession.ts (hunk 8 above)
Comment (dhawal1) [reply ts/packages/core/src/models/ToolRouterSession.ts:343]: done

Comment (dhawal1) [reply ts/packages/core/src/models/ToolRouterSession.ts:343]: Done. Extracted `findLocalTool()` and `executeLocalTool()` into `localToolExecution.ts`. The `routeMultiExecute` orchestration stays in `ToolRouterSession` since it coordinates both local and remote.

This extraction was also needed for the sibling local tool routing fix — `SessionContextImpl` now shares these standalone functions to route sibling calls in-process.

**Comment by jkomyno** [comment]
## Code Review — PR #2877

**Reviewer:** Claude Code (6 parallel agents: Security, Performance, Architecture, Patterns, Simplicity, TypeScript Quality)

---

### Findings Summary

- **Total:** 20 findings
- **🔴 P1 CRITICAL:** 2 — should fix before merge
- **🟡 P2 IMPORTANT:** 8 — should fix
- **🔵 P3 NICE-TO-HAVE:** 10 — enhancements

---

### 🔴 P1 — CRITICAL (Blocks Merge)

#### 1. Unbounded recursive stack overflow in `buildSessionContext()`
**File:** `ToolRouterSession.ts:343-354` | **Found by:** ALL 6 agents

The `tryLocalExecute` callback calls `this.buildSessionContext()` recursively. If a custom tool calls itself or two tools call each other, this produces infinite recursion → **process crash**. User-supplied code makes this trivially triggerable:

```typescript
const evil = createCustomTool({
  slug: 'EVIL', name: 'Evil', description: 'Boom',
  inputParams: z.object({}),
  execute: async (_input, session) => session.execute('EVIL', {}), // 💥 stack overflow
});
```

**Fix:** Add a depth counter (max ~10) to `buildSessionContext(depth)`. Return an error response when exceeded.

#### 2. Custom tool slugs can shadow `COMPOSIO_` meta tools and any remote tool
**File:** `customTool.types.ts:101-108` | **Found by:** Security Sentinel

The slug validation only rejects `LOCAL_` prefix. A custom tool with slug `COMPOSIO_MULTI_EXECUTE_TOOL` will shadow the meta tool in `session.execute()`. More broadly, **any** custom tool slug matching a remote tool silently shadows it — `findCustomTool` is always checked first in `execute()` (line 312).

**Fix:** Add `COMPOSIO_` to the rejected prefix list in `CreateCustomToolBaseSchema`. Consider a warning when slugs match remote tools.

---

### 🟡 P2 — IMPORTANT (Should Fix)

#### 3. `proxyExecute` stub on public `SessionContext` interface
**Files:** `customTool.types.ts:77-82`, `SessionContext.ts:74-79` | **Found by:** 4 agents

Always throws. Violates Interface Segregation Principle. Users see it in autocomplete → runtime error. Since this is experimental, removing is non-breaking.

#### 4. `z.custom<CustomTool>()` performs no runtime validation
**File:** `toolRouter.types.ts:220` | **Found by:** Security Sentinel

`z.custom()` without a validator accepts any value at runtime. A malformed object missing `inputParams` or `execute` won't be caught until deep in execution.

**Fix:** Add structural validation in `buildCustomToolsMap()` or replace with a proper Zod schema.

#### 5. Plain `Error` instead of codebase error classes
**Files:** `CustomTool.ts`, `ToolRouterSession.ts` | **Found by:** Pattern Recognition

The codebase uses `ValidationError`, `ComposioProviderNotDefinedError` everywhere else. This PR uses plain `Error` for all throws.

#### 6. Per-tool `buildSessionContext()` allocation in `routeMultiExecute`
**File:** `ToolRouterSession.ts:413` | **Found by:** Performance Oracle

Each local tool in a batch creates its own identical `SessionContextImpl`. These are functionally equivalent and stateless. Hoist `const ctx = this.buildSessionContext()` above the `.map()`.

#### 7. Duplicated local-tool dispatch logic
**File:** `ToolRouterSession.ts:86-98` vs `183-200` | **Found by:** Architecture, Simplicity

Nearly identical slug-lookup-and-execute code in `tools()` routing and `localTools()` execute callback. Extract a shared `private dispatchLocalTool(slug, args)` method.

#### 8. `use()` returns Session with non-functional `localTools()`
**File:** `ToolRouter.ts:168-176`, `toolRouter.types.ts:439` | **Found by:** Architecture Strategist

`localTools` is required on the `Session` interface, but sessions from `use()` get no `customToolsMap`. `localTools()` always throws on those sessions, yet the type promises it works.

**Fix:** Make `localTools` optional on the Session interface, or narrow `use()` return type.

#### 9. Mixed "local" vs "custom" naming in public API
**Found by:** Architecture Strategist

Users call `createCustomTool()`, pass `customTools` in config, but then call `session.localTools()`. Before graduating from experimental, consider renaming `localTools()` → `customTools()`.

#### 10. `this.userId!` non-null assertion relies on implicit invariant
**File:** `ToolRouterSession.ts:346` | **Found by:** TypeScript Quality

The `!` relies on call-site discipline across methods. Add an explicit guard at the start of `buildSessionContext()`.

---

### 🔵 P3 — NICE-TO-HAVE

| # | Finding | File |
|---|---------|------|
| 11 | `customToolExecution.ts` should be PascalCase to match other model files | `models/` |
| 12 | Experimental types leaked via `export *` in main `index.ts` | `index.ts:15` |
| 13 | Multi-execute merge doesn't preserve input ordering (`[...remote, ...local]`) | `ToolRouterSession.ts:457` |
| 14 | `zodToJsonSchema` output cast without runtime guard | `CustomTool.ts:108-111` |
| 15 | No concurrency limit on parallel local tool execution | `ToolRouterSession.ts:411` |
| 16 | Error messages may leak sensitive details from user execute fns | `customToolExecution.ts:55` |
| 17 | Missing tests: circular recursion, `execute` returning undefined, non-array `tools` | `test/` |
| 18 | Constructor has 7 positional params — consider options object | `ToolRouterSession.ts:45` |
| 19 | `logId: 'local'` sentinel value undocumented | `ToolRouterSession.ts:318` |
| 20 | `successful` re-derived from `!error` instead of forwarding backend field | `SessionContext.ts:63` |

---

### Verdict

**The feature design and implementation are solid.** The factory pattern, dual-map routing, parallel local/remote execution, and test coverage (1374 lines across unit + E2E) are all well-executed. The two P1 items (stack overflow + slug shadowing) are the only merge blockers — both are small fixes (~15 lines total). The P2 items are quality improvements to address before graduating from experimental.

*Review performed by Claude Code with 6 specialized agents: Security Sentinel, Performance Oracle, Architecture Strategist, Pattern Recognition Specialist, Code Simplicity Reviewer, and TypeScript Quality Reviewer.*

File: ts/packages/core/src/models/ToolRouterSession.ts (hunk 7 above)
Comment (dhawal1) [reply ts/packages/core/src/models/ToolRouterSession.ts]: Resolved — `localTools()` has been removed entirely. Custom tools are now automatically routed via `COMPOSIO_MULTI_EXECUTE_TOOL` interception in `session.tools()` for agent/SDK flows, and indexed by the backend for MCP search flows. No separate injection needed.

File: ts/packages/core/src/types/customTool.types.ts (hunk 9 above)
Comment (dhawal1) [reply ts/packages/core/src/types/customTool.types.ts]: Now `extendsToolkit` — aligned with backend `extends_toolkit` field from hermes#8614. The name conveys that this tool *extends* an existing Composio toolkit (inheriting its auth), rather than simply being associated with one.

File: ts/packages/core/src/models/CustomTool.ts (hunk 6 above)
Comment (dhawal1) [reply ts/packages/core/src/models/CustomTool.ts]: Now `extendsToolkit` — see reply on the types file comment.

File: ts/packages/core/src/models/ToolRouterSession.ts (hunk 8 above)
Comment (dhawal1) [reply ts/packages/core/src/models/ToolRouterSession.ts:343]: Done — extracted to `customToolExecution.ts` (`findCustomTool`, `executeCustomTool`), `SessionContext.ts`, and `proxyParamsTransform.ts`. The multi-execute orchestration stays in `ToolRouterSession` since it coordinates both local and remote execution paths.

**Comment by dhawal1** [comment]
$(cat << 'REPLY'
## Response to code review

Thanks for the thorough review @jkomyno! Here's the status of each finding after the latest refactor (aligned with backend hermes#8614):

### 🔴 P1 — CRITICAL

**1. Stack overflow in `buildSessionContext()` recursion** — Acknowledged, will fix in follow-up. The plan is to eliminate the recursive callback pattern entirely: give `SessionContextImpl` a direct reference to `customToolsMap` and have `execute()` do the lookup inline, passing `this` (same context) to sibling tools. No recursion, no depth counter needed.

**2. Custom tool slugs shadow `COMPOSIO_` meta tools** — Will add `COMPOSIO_` to the rejected prefix list in `CustomToolSlugSchema`. Small fix.

### 🟡 P2 — IMPORTANT

| # | Finding | Status |
|---|---|---|
| 3 | `proxyExecute` stub | **Fixed** — fully implemented with backend session proxy endpoint |
| 4 | `z.custom<CustomTool>()` no runtime validation | Deferred — structural validation happens downstream in `buildCustomToolsMap()` when accessing properties. Adding upfront validation would duplicate the checks. |
| 5 | Plain `Error` instead of codebase error classes | Deferred — will address when graduating from experimental |
| 6 | Per-tool `buildSessionContext()` allocation | Will be resolved by P1 #1 fix (single context instance) |
| 7 | Duplicated local-tool dispatch | **Fixed** — extracted to `customToolExecution.ts` |
| 8 | `use()` returns Session with non-functional `localTools()` | **Fixed** — `localTools()` removed entirely |
| 9 | Mixed "local" vs "custom" naming | **Fixed** — standardized to "custom" in public API. Internal `LOCAL_` prefix is backend's naming convention. |
| 10 | `this.userId!` non-null assertion | Guarded by constructor check: `if (customToolsMap && !userId) throw` |

### 🔵 P3 — NICE-TO-HAVE

Most are style/design preferences that can be addressed incrementally. #12 (experimental types leaked via `export *`) is fixed — we now have explicit exports.
REPLY
)

**Comment by dhawal1** [comment]
@haxzie re: "What's the difference between customTool and localTool? Can we unify the names?"

Resolved — unified to "custom tool" everywhere in the public API:
- `experimental_createTool()` / `experimental_createToolkit()` (top-level exports)
- `createCustomTool()` / `createCustomToolkit()` (from `@composio/core/experimental`)
- `experimental.customTools` / `experimental.customToolkits` in session config
- Types: `CustomTool`, `CustomToolkit`, `CreateCustomToolParams`, `CreateCustomToolkitParams`

The `LOCAL_` prefix is only used internally for the backend slug format (`LOCAL_<SLUG>`), never exposed in the user-facing API. The old `localTools()` method has been removed entirely.

File: ts/docs/api/tool-router.md [hunk 11]
```diff
@@ -347,6 +430,48 @@ Meta tools allow you to:
...
+  console.log(result.primaryToolSlugs); // e.g. ['GMAIL_SEND_EMAIL']
+}
+```
+
+### `proxyExecute()`
+
+Proxy an API call through Composio's auth layer using the session's connected account. The backend resolves the connected account from the toolkit within the session.
+
+```typescript
+const result = await session.proxyExecute({
```
Comment (dhawal1) [inline ts/docs/api/tool-router.md:465]: blocked on staneless sdk to get correct types

**Review by haxzie — APPROVED** [review]
LGTM 👍🏽

**Comment by jkomyno** [comment]
Not LGTM, the newly introduced e2e tests are failing: `@e2e-tests/node-custom-tools:test:e2e:node`
